Source/DafnyLanguageServer/DafnyCodeActions.cs

```diff
@@ -33,22 +33,145 @@ public class DafnyCodeActions {
     var edit = new DafnyCodeActionEdit[] { new DafnyCodeActionEdit(range, "") };
     var action = new InstantDafnyCodeAction(title, new List<Diagnostic> { diagnostic }, edit);
     return new List<DafnyCodeAction> { action };
   }
 
+  public static List<DafnyCodeAction> RemoveAction(IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) {
+    string toBeRemoved = input.Extract(range).Trim();
+    string title = "remove '" + toBeRemoved + "'";
+    var edit = new DafnyCodeActionEdit[] { new DafnyCodeActionEdit(range, "") };
+    var action = new InstantDafnyCodeAction(title, new List<Diagnostic> { diagnostic }, edit);
+    return new List<DafnyCodeAction> { action };
+  }
+
+  public static void Add(ErrorID errorID, ActionSignature action) {
+    codeActionMap.Add(errorID, action);
+  }
+
+  public static void AddRemoveAction(ErrorID errorID, String title) {
+    Add(errorID, (IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) => RemoveAction(title, diagnostic, range));
+  }
+
+  // Default title is "remove'X'" where X is the content of the range
+  public static void AddRemoveAction(ErrorID errorID) {
+    Add(errorID, (IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) => RemoveAction(input, diagnostic, range));
+  }
+
+  public static void AddInsertAction(ErrorID errorID, String title, String newContent) {
+    Add(errorID, (IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) =>
+      InsertAction(title, diagnostic, range, newContent));
+  }
+
+  public static void AddReplaceAction(ErrorID errorID, String title, String newContent) {
+    Add(errorID, (IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) => ReplacementAction(title, diagnostic, range, newContent));
+  }
+
+  public static void AddReplaceAction(ErrorID errorID, String newContent) {
+    Add(errorID, (IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) => ReplacementAction(input, diagnostic, range, newContent));
+  }
+
+  public static void AddReplacementActions(ErrorID errorID, Tuple<String, String>[] actions) {
+    codeActionMap.Add(errorID, (IDafnyCodeActionInput input, Diagnostic diagnostic, Range range) => ReplacementActions(diagnostic, range, actions));
+  }
+
   private static bool initialized = false;
   public static void init() {
     if (initialized) {
       return;
     }
     initialized = true;
-    codeActionMap.Add(ErrorID.p_bad_const_initialize_op, (Diagnostic diagnostic, Range range) => ReplacementAction("replace = with :=", diagnostic, range, ":="));
-    codeActionMap.Add(ErrorID.p_abstract_not_allowed, (Diagnostic diagnostic, Range range) => RemoveAction("remove 'abstract'", diagnostic, range));
-    codeActionMap.Add(ErrorID.p_no_leading_underscore, (Diagnostic diagnostic, Range range) => RemoveAction("remove underscore", diagnostic, range));
+    AddRemoveAction(ErrorID.p_duplicate_modifier, "remove duplicate modifier'");
```

Comment:
Please register the actions where the error message is created.

Reply:
The problem is that registering and using the actions is done in the language server. The error message is created in the DafnyCore code. We did not want to drage LS code into the DafnyCore

Reply:
I think refactorings can be part of DafnyCore, and it's not something that should only be available to do language server. You can imagine running `dafny format --include-code-actions` and also applying these code actions.